src/VisualStudio/Core/Next/Remote/ServiceHubRemoteHostClient.cs

```diff
@@ -36,23 +36,14 @@ internal sealed partial class ServiceHubRemoteHostClient : RemoteHostClient
         {
             using (Logger.LogBlock(FunctionId.ServiceHubRemoteHostClient_CreateAsync, cancellationToken))
             {
-                // let each client to have unique id so that we can distinguish different clients when service is restarted
-                var currentInstanceId = Interlocked.Add(ref s_instanceId, 1);
-
                 var primary = new HubClient("ManagedLanguage.IDE.RemoteHostClient");
-                var current = $"VS ({Process.GetCurrentProcess().Id}) ({currentInstanceId})";
-
-                var hostGroup = new HostGroup(current);
                 var timeout = TimeSpan.FromMilliseconds(workspace.Options.GetOption(RemoteHostOptions.RequestServiceTimeoutInMS));
-                var remoteHostStream = await RequestServiceAsync(primary, WellKnownRemoteHostServices.RemoteHostService, hostGroup, timeout, cancellationToken).ConfigureAwait(false);
-
-                var instance = new ServiceHubRemoteHostClient(workspace, primary, hostGroup, remoteHostStream);
-
-                // make sure connection is done right
-                var host = await instance._rpc.InvokeAsync<string>(nameof(IRemoteHostService.Connect), current, TelemetryService.DefaultSession.SerializeSettings()).ConfigureAwait(false);
 
-                // TODO: change this to non fatal watson and make VS to use inproc implementation
-                Contract.ThrowIfFalse(host == current.ToString());
+                // Retry (with timeout) until we can connect to RemoteHost (service hub process). 
+                // we are seeing cases where we failed to connect to service hub process when a machine is under heavy load.
+                // (see https://devdiv.visualstudio.com/DevDiv/_workitems/edit/481103 as one of example)
+                var instance = await RetryRemoteCallAsync<(Workspace workspace, HubClient primary, TimeSpan timeout), IOException, ServiceHubRemoteHostClient>(
```

Comment:
Are all these typ eparameter really necessary?

Reply:
I tried not to capture anything. if we are okay with capturing, then I dont need all these.